runtime/oti/ObjectHash.hpp

```diff
@@ -203,6 +203,23 @@ class VM_ObjectHash
 		return inlineConvertValueToHash(vm, (UDATA)objectPointer);
 	}
 
+	/**
+	 * Fetch objectPointer's hashcode when objectPointer is a contiguous object
+	 *
+	 * @pre objectPointer must be a valid object reference
+	 *
+	 * @param vm			a java VM
+	 * @param objectPointer 	a valid object reference.
+	 * @param objectClass 	class of objectPointer
+	 * @param offset 	objectPointer size offset
+	 */
+	static VMINLINE I_32
+	inlineContiguousObjectHashCode(J9JavaVM *vm, j9object_t objectPointer, J9Class *objectClass, UDATA offset)
+	{
+		J9ROMArrayClass *romClass = (J9ROMArrayClass*)objectClass->romClass;
+		offset = ROUND_UP_TO_POWEROF2((offset << (romClass->arrayShape & 0x0000FFFF)) + vm->contiguousHeaderSize, sizeof(I_32));
+		return *(I_32*)((UDATA)objectPointer + offset);
+	}
```

Comment:
You could introduce 2 more inline helpers inlineObjectHashCodeFull/Compressed. Nothing to share just to make the heavily nested 'if' easier to read. It's subjective if it's helping or not (one may find moving eyes to another spot in the code to follow the flow more annoying) - up to you if you want to do it or not.

Reply:
Ok, yes I see both sides of the argument sure. I decided to add them, and have made sure to put the helpers right above so there is minimal eye movement needed.